### PR TITLE
HDDS-4174. Add current HDDS layout version to Datanode heartbeat/registration

### PR DESCRIPTION
**Pull request by prashantpogde**
## What changes were proposed in this pull request?

Add current HDDS layout version to DataNode heartbeat/registration

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4174

## How was this patch tested?
Successful Build

### CODE REVIEW

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto [hunk 1]
```diff
@@ -390,6 +404,14 @@ message ClosePipelineCommandProto {
   required int64 cmdId = 2;
 }
 
+/**
+ * This command asks the DataNode to finalize a new HDDS layout version.
+ */
+message FinalizeNewLayoutVersionCommandProto {
+  required bool finalizeNewHDDSLayoutVersion = 1 [default = false];
```
Comment (elek) [inline hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:411]: What is the use case the command where this flag=false. Is it a no-op request?

**Review by elek — APPROVED** [review]
Thanks the patch @prashantpogde 

LGTM. Only one question inline.

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto [hunk 2]
```diff
@@ -92,6 +101,7 @@ message SCMRegisterRequestProto {
   required NodeReportProto nodeReport = 2;
   required ContainerReportsProto containerReport = 3;
   required PipelineReportsProto pipelineReports = 4;
+  optional HDDSLayoutVersionProto dataNodeLayoutVersion = 5;
```
Comment (linyiqun) [inline hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:104]: Why not this be a required field? when Datanode firstly register on SCM, it should have its own HDDS layout version.

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto [hunk 3]
```diff
@@ -303,6 +315,8 @@ message SCMCommandProto {
   optional ReplicateContainerCommandProto replicateContainerCommandProto = 6;
   optional CreatePipelineCommandProto createPipelineCommandProto = 7;
   optional ClosePipelineCommandProto closePipelineCommandProto = 8;
+  optional FinalizeNewLayoutVersionCommandProto
+  finalizeNewLayoutVersionCommandProto = 9;
```
Comment (amaliujia) [inline hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:319]: Why not use a single line: `optional FinalizeNewLayoutVersionCommandProto=9`?

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto (hunk 1 above)
Comment (prashantpogde) [reply hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:411]: yes, that would be a no op request.

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto (hunk 2 above)
Comment (prashantpogde) [reply hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:104]: Datanode need not pass the HDDSLaoutVersionProto information during registration.

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto (hunk 3 above)
Comment (prashantpogde) [reply hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:319]: > 80 columns

Comment (amaliujia) [reply hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:319]: Ah I see. Didn't know Ozone also enforce that in proto files (very common that message name is very long).

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto (hunk 2 above)
Comment (prashantpogde) [reply hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto:104]: All new fields added should be optional for backward compatibility.

**Review by avijayanhwx — APPROVED** [review]
Thanks for working on this @prashantpogde, and for the reviews @elek, @linyiqun & @amaliujia.

File: hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto (hunk 1 above)
Comment (elek) [reply hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto]: Is there any use case for this noop request? Will it be used by somebody? Why do we need this field?

**Pull request merged**